File: javascript/phone-number/phone-number.js
const INVALID_NUMBER = "0000000000";

function PhoneNumber (input) {
  this.num = this.validateInput(input);
}

PhoneNumber.prototype = {
  constructor: PhoneNumber,

  areaCode: function () {
    if (!this.num) { return INVALID_NUMBER; }

    return this.num.substr(0,3);
  },

  number: function () {
    if (!this.num) { return INVALID_NUMBER; }

    return this.num;
  },

  toString: function () {
    if (!this.num) { return INVALID_NUMBER; }

    return this.num.replace(/^(\d{3})(\d{3})(\d{4})/, "($1) $2-$3");
  },

  validateInput: function (input) {
    var digits = input.replace(/\D/g, '');

    if (digits.length < 10) { return null; }
    if (digits.length === 11) {
      if (digits[0] == 1) { digits = digits.substr(1); }
      else { return null; }
    }
    if (digits.length > 11) { return null; }

    return digits;
  }
}

module.exports = PhoneNumber;

File: ruby/phone-number/phone_number.rb
class PhoneNumber
  INVALID_NUMBER = '0000000000'
  JUST_TEN_DIGITS = /\A\d{10}\z/

  def initialize(number)
    @number = _clean_number(number)
  end

  attr_reader :number

  @property
  def area_code
    number[0, 3]
  end

  @property
  def exchange_number
    number[3, 3]
  end

  @property
  def subscriber_number
    number[6, 4]
  end

  def to_s
    format "(%03d) %03d-%04d", area_code, exchange_number, subscriber_number
  end

  private

    def _clean_number(str)
      digits = _remove_punct_and_space(str)
      digits = _remove_leading_1(digits)
      return INVALID_NUMBER if digits !~ JUST_TEN_DIGITS

      digits
    end

    def _remove_leading_1(str)
      if str.size == 11 && str[0] == '1'
        return str[1..str.size]
      else
        return str
      end
    end

    def _remove_punct_and_space(str)
      str.gsub(/[[:punct:][:space:]]/, '')
    end
end